### ISSUE TITLE
Index for TiFlash

### ISSUE THREAD

**Issue by phamtai97** [pingcap/tidb#33022]
Hi all,
I am running performance for TiFlash and Clickhouse to compare them. The Clickhouse can order and create the index for the table, so it can perform the SQL fastly.  But when I use the TiFlash, it does not support creating the index. I think that the performance won't be as good as Clickhouse.

I have questions:

- The data of TiFlash is replicated from TiKV,  How to the index of TikV hosted on TiFlash? 
- The TiFlash is good for scanning data or count or sum by column (column is created index by TiKV)? Because I explain the SQL query, the result is TiKV. For example:
<img width="897" alt="Screen Shot 2022-03-13 at 20 33 06" src="https://user-images.githubusercontent.com/38773351/158061991-8b0da513-b7cd-4dc5-a800-306b3a1d8d31.png">
<img width="800" alt="Screen Shot 2022-03-13 at 20 31 49" src="https://user-images.githubusercontent.com/38773351/158062026-87cc588d-d51c-4ec6-bd34-fec90d09d981.png">

**Comment by phamtai97** [comment]
Can you help me? @tiancaiamao

**Comment by tiancaiamao** [comment]
1. The data of TiFlash is replicated from TiKV, How to the index of TikV hosted on TiFlash?

TiFlash can't use TiKV's index... they're two different storage engine.
TiDB is the shared computing layer, it'll decide whether to send the query to TiFlash or TiKV
But it can't do something like "query the index from TiKV, then use the index result to fetch data from TiFlash"
It checks something like "this query is scan heavy, use TiFlash would help, so just choose TiFlash ... this query is a very simple OLTP query, TiKV is better for this task, so just choose TiKV"

2. The TiFlash is good for scanning data or count or sum by column

Exactly!
TiFlash is a column-oriented storage engine, while TiKV is row-oriented.
It means for "select count(c) from t", TiKV need to fetch every row and cut the 'c' column, the read amplification is high.

**Comment by phamtai97** [comment]
Thank you for your answer.
> 1. The data of TiFlash is replicated from TiKV, How to the index of TikV hosted on TiFlash?
TiFlash can't use TiKV's index... they're two different storage engine.
TiDB is the shared computing layer, it'll decide whether to send the query to TiFlash or TiKV
But it can't do something like "query the index from TiKV, then use the index result to fetch data from TiFlash"
It checks something like "this query is scan heavy, use TiFlash would help, so just choose TiFlash ... this query is a very simple OLTP query, TiKV is better for this task, so just choose TiKV"

=> Currently, the TiFlash does not support query index => It affects the performance of query? In the future, Are you planning to implement an index for TiFlash?

> 2. The TiFlash is good for scanning data or count or sum by column
Exactly!
TiFlash is a column-oriented storage engine, while TiKV is row-oriented.
It means for "select count(c) from t", TiKV need to fetch every row and cut the 'c' column, the read amplification is high.

=> But when I explain the count query, the tidb chooses the TiKV to perform this query? 
<img width="800" alt="Screen Shot 2022-03-13 at 20 31 49" src="https://user-images.githubusercontent.com/38773351/158062026-87cc588d-d51c-4ec6-bd34-fec90d09d981.png">

**Comment by phamtai97** [comment]
Can you help me? @tiancaiamao

**Comment by tiancaiamao** [comment]
> In the future, Are you planning to implement an index for TiFlash?

If you mean get index from TiKV and then use the index in TiFlash, no... the organization of data in the two engine is totally different and this is impossible...

> But when I explain the count query, the tidb chooses the TiKV to perform this query?

Because `where account_id = ?` can use TiKV index, so it's fast to run the query in TiKV, so TiDB choose the TiKV storage engine.
If this query is execute in TiFlash, the need to scan a lot of data and filter out the `where account_id = ?`  condition to get the final result.

> Do you have a comparison test between TiFlash and Clickhouse?

I don't know much about Clickhouse, and do I know much about TiFlash, neither.

**Comment by phamtai97** [comment]
>If you mean get index from TiKV and then use the index in TiFlash, no... the organization of data in the two engine is totally different and this is impossible.

No, I mean is TiFlash indexed in the future. You can index of the Clickhouse: https://clickhouse.com/docs/en/engines/table-engines/mergetree-family/mergetree/#available-types-of-ind

**Comment by tiancaiamao** [comment]
I'm not familiar with the TiFlash roadmap, sorry~
PTAL @flowbehappy

**Comment by flowbehappy** [comment]
> Hi all, I am running performance for TiFlash and Clickhouse to compare them. The Clickhouse can order and create the index for the table, so it can perform the SQL fastly. But when I use the TiFlash, it does not support creating the index. I think that the performance won't be as good as Clickhouse.
> 
> I have questions:
> 
> * The data of TiFlash is replicated from TiKV,  How to the index of TikV hosted on TiFlash?
TiFlash only reuses the PK index, i.e. data is sorted by Primary Key. Other indexes like secondary indexes created by TiKV are not used by TiFlash. However, TiFlash automatically creates its own kinds of indexes which are more suitable for AP workloads. Currently, we support Min-Max Index for Integer and DateTime datatypes. And there are plans for more index types in the future.

> * The TiFlash is good for scanning data or count or sum by column (column is created index by TiKV)? Because I explain the SQL query, the result is TiKV. For example:
> 

Yes. TiFlash is playing the  "AP" role of HTAP in TiDB databases. So it is designed for running fast scan and fast aggregation like sum. TiFlash synchronizes table schema modifications in realtime. 

About your example, did you create TiFlash replicas? TiFlash replicas should be created manually. Check this document for details:

https://docs.pingcap.com/tidb/v6.0/use-tiflash
